portablemc/optifine.py

```diff
@@ -0,0 +1,721 @@
+            length = struct.unpack(">H", patch_stream.read(2))[0]
+            output_stream.write(append(length, patch_stream))
+        elif command == 248:
+            # Append `unsigned int` bytes of data
+            length = struct.unpack(">I", patch_stream.read(4))[0]
+            output_stream.write(append(length, patch_stream))
+        elif command == 249:
+            # Copy `unsigned byte` bytes from `unsigned short` offset
+            offset = struct.unpack(">H", patch_stream.read(2))[0]
+            length = struct.unpack("B", patch_stream.read(1))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 250:
+            # Copy `unsigned short` bytes from `unsigned short` offset
+            offset = struct.unpack(">H", patch_stream.read(2))[0]
+            length = struct.unpack(">H", patch_stream.read(2))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 251:
+            # Copy `unsigned int` bytes from `unsigned short` offset
+            offset = struct.unpack(">H", patch_stream.read(2))[0]
+            length = struct.unpack(">I", patch_stream.read(4))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 252:
+            # Copy `unsigned byte` bytes from `unsigned int` offset
+            offset = struct.unpack(">I", patch_stream.read(4))[0]
+            length = struct.unpack("B", patch_stream.read(1))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 253:
+            # Copy `unsigned short` bytes from `unsigned int` offset
+            offset = struct.unpack(">I", patch_stream.read(4))[0]
+            length = struct.unpack(">H", patch_stream.read(2))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 254:
+            # Copy `unsigned int` bytes from `unsigned int` offset
+            offset = struct.unpack(">I", patch_stream.read(4))[0]
+            length = struct.unpack(">I", patch_stream.read(4))[0]
+            output_stream.write(copy(offset, length))
+        elif command == 255:
+            # Copy `unsigned int` bytes from `long` offset
+            offset = struct.unpack(">Q", patch_stream.read(8))[0]
+            length = struct.unpack(">I", patch_stream.read(4))[0]
+            output_stream.write(copy(offset, length))
+        else:
+            # Unrecognized command; append the command byte itself
+            output_stream.write(bytes([command]))
+
+    return output_stream.getvalue()
+
+
+class OptifineStartInstallEvent:
+    __slots__ = tuple()
+
+
+class OptifinePatchEvent:
+    __slots__ = "location","total","done"
+    def __init__(self, location: str,total: int,done: int):
+        self.location = location
+        self.total = total
+        self.done = done
+
+
+class OptifineEndInstallEvent:
+    __slots__ = "version",
+    def __init__(self,version: str):
+        self.version = version
+
+
+class Patcher:
```

Comment:
If this class is internal, please prefix it with an underscore, and in general please do that for everything that is internal! We want to be very precise about what is publicly exposed, to ensure backward compatibility.